File: 0941-sort-array-by-parity/0941-sort-array-by-parity.py
class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        even, odd = 0, len(nums)-1
        while even < odd:
            if nums[even] % 2 != 0:
                # replace with even from odd until we fine odd
                while even < odd:
                    if nums[odd] % 2 != 0:
                        odd -= 1
                    else:
                        break;
                #swap number
                nums[even], nums[odd] = nums[odd], nums[even]
                odd -= 1
            else:
                even += 1
        return nums
        
        

File: 0941-sort-array-by-parity/0941-sort-array-by-parity.java
class Solution {
    public int[] sortArrayByParity(int[] nums) {
        int left = 0, right = nums.length-1;
        while(left < right){
            //traverse left
            while(left < right && nums[left] % 2 == 0){
                left++;
            }

            //traverse right
            while(left < right && nums[right] % 2 != 0){
                right--;
            }
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;            
        }
        return nums;
    }
}